File: src/main/java/edu/zjnu/designpattern/zhaihongwei/singleton/singleton/threadSafetyVolatile/Singleton.java
package edu.zjnu.designpattern.zhaihongwei.singleton.singleton.threadSafetyVolatile;

/**
 * 高效的线程安全的单例模式实现类
 *
 * @author 杨海波
 */
public class Singleton {
    private static volatile Singleton instance;

    private Singleton() {
    }

    public static Singleton getInstance() {
        if (instance == null) {
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();
                }
            }
        }

        return instance;
    }

}

File: Sundry/_design_pattern/Go/_singleton/singleton.go
package singleton

import "sync"

// Singleton 是单例模式接口，导出的
// 通过该接口可以避免 GetInstance 返回一个包私有类型的指针
type Singleton interface {
	foo()
}

// singleton 是单例模式类，包私有的
type singleton struct{}

func (s singleton) foo() {}

var (
	instance *singleton
	once     sync.Once
)

// GetInstance 用于获取单例模式对象
func GetInstance() Singleton {
	once.Do(func() {
		instance = &singleton{}
	})

	return instance
}
